Retry apphost creation.

peterhuene opened the pull request:
This PR implements a parameterized retry count for creating the apphost.

Like the `Copy` task from MSBuild, the `CreateAppHost` task now takes a
parameter to specify the number of retries and the delay between retries (in
milliseconds) to perform if the creation fails.

This should help alleviate build failures when external processes are locking
the intermediate apphost during a build (especially on Windows).

Fixes [AB#950462](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/950462).
Fixes dotnet/cli#11650.

Commit:
Retry apphost creation.

This commit implements a parameterized retry count for creating the apphost.

Like the `Copy` task from MSBuild, the `CreateAppHost` task now takes a
parameter to specify the number of retries and the delay between retries (in
milliseconds) to perform if the creation fails.

This should help alleviate build failures when external processes are locking
the intermediate apphost during a build (especially on Windows).

Fixes devdiv#950462.
Fixes dotnet/cli#11650.

peterhuene commented on src/Tasks/Microsoft.NET.Build.Tasks/CreateAppHost.cs:
```diff
@@ -28,34 +35,63 @@ public class CreateAppHost : TaskBase
                     }
+                    catch (Exception ex) when (ex is IOException ||
+                                               ex is UnauthorizedAccessException)
+                                               //ex is ResourceUpdater.HResultException)
```
This exception is currently private in core-setup.  We would need to update the `ResourceUpdater` to be able to catch it more specifically than `Exception`.

nguerrera replied:
Maybe `ex.GetType().Name == “HResultException”` for now against a bug on core-setup or were you going to PR the fix to core-setup and wait for it to flow. This is the case we’re hitting right? 

Also my suggestion would be for core-setup to make the public exception contract more abstract like ResourceUpdateFailedException or something. They could still include HResult in Exception.HResult property and message. This would allow the API to become x-plat without needing changes here.

peterhuene replied:
I can live with `ex.GetType().Name` check for now and file a core-setup bug to make the exception public (and better).

peterhuene replied:
Files as https://github.com/dotnet/core-setup/issues/7516.

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -355,12 +355,18 @@ Copyright (c) .NET Foundation. All rights reserved.
     <PropertyGroup>
       <_UseWindowsGraphicalUserInterface Condition="($(RuntimeIdentifier.StartsWith('win')) or $(DefaultAppHostRuntimeIdentifier.StartsWith('win'))) and '$(OutputType)'=='WinExe'">true</_UseWindowsGraphicalUserInterface>
+      <CreateAppHostRetries Condition="'$(CreateAppHostRetries)' == ''">3</CreateAppHostRetries>
+      <CreateAppHostRetryDelay Condition="'$(CreateAppHostRetryDelay)' == ''">1000</CreateAppHostRetryDelay>
```
My instinct is to use `$(CopyRetryCount)` and `$(CopyRetryDelayMilliseconds)`--this feels very analogous. Or maybe as the defaults for the distinct properties?

Also, nit: usually this type of default property is defined outside a target.

nguerrera replied:
I think either reusing completely or reusing defaults would be fine.

peterhuene replied:
I'm fine with reusing the same properties.  I'll fix.

peterhuene replied:
Where do `CopyRetryCount` and `CopyRetryDelayMilliseconds` get defaulted?

I think we should be defaulting the apphost creation retry count to non-zero.

peterhuene replied:
I moved the defaulting of the properties out of the target and will reuse `CopyRetryCount` and `CopyRetryDelayMilliseconds` if I can find where they would be defaulted (and the defaults make sense for this use case).

nguerrera replied:
Looks like the values are assumed to be empty by default, and the copy task code is what defaults them. Good catch.

We could still reuse the same variables, but we would have to copy the defaults into the task. Maybe that's ok?

https://github.com/microsoft/msbuild/blob/acc2569a30fc9749942f4cdc58811e12acabf096/src/Tasks/Copy.cs#L76
https://github.com/microsoft/msbuild/blob/acc2569a30fc9749942f4cdc58811e12acabf096/src/Tasks/Copy.cs#L89

peterhuene replied:
That approach works for me.  I'll do that.

Commit:
Code review feedback changes.

peterhuene commented:
Hmm, no idea why the new test is failing on the Windows full framework legs (it's printing four warnings instead of the expected two that every other platform is getting).  I'll have to try to reproduce.

Commit:
Fix apphost test for full framework CI.

When using the full-framework MSBuild, the test output contains the warnings
again in the summary, so account for this in the expected number of warnings in
the test.
